pkg/controller/seed-controller-manager/mla/userprojectbinding_controller.go

```diff
@@ -0,0 +1,250 @@
+			return reconcile.Result{}, fmt.Errorf("unable to update grafana user role: %w (status: %s, message: %s)", err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *userProjectBindingReconciler) getOrgByProjectID(ctx context.Context, projectID string) (grafanasdk.Org, error) {
+	project := &kubermaticv1.Project{}
+	if err := r.Get(ctx, types.NamespacedName{Name: projectID}, project); err != nil {
+		return grafanasdk.Org{}, fmt.Errorf("failed to get project: %w", err)
+	}
+
+	orgID, ok := project.GetAnnotations()[grafanaOrgAnnotationKey]
+	if !ok {
+		return grafanasdk.Org{}, fmt.Errorf("project should have grafana org annotation set")
+	}
+	id, err := strconv.ParseUint(orgID, 10, 32)
+	if err != nil {
+		return grafanasdk.Org{}, err
+	}
+	return r.grafanaClient.GetOrgById(ctx, uint(id))
+}
+
+func (r *userProjectBindingReconciler) handleDeletion(ctx context.Context, userProjectBinding *kubermaticv1.UserProjectBinding) error {
+	org, err := r.getOrgByProjectID(ctx, userProjectBinding.Spec.ProjectID)
+	if err != nil {
+		return err
+	}
+	user, err := r.getGrafanaOrgUser(ctx, org.ID, userProjectBinding.Spec.UserEmail)
+	if err != nil {
+		return fmt.Errorf("unable to get user : %w", err)
+	}
+	if user != nil {
+		status, err := r.grafanaClient.DeleteOrgUser(ctx, user.OrgId, user.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete org user: %w (status: %s, message: %s)", err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+		}
+	}
+
+	kubernetes.RemoveFinalizer(userProjectBinding, mlaFinalizer)
+	if err := r.Update(ctx, userProjectBinding); err != nil {
+		return fmt.Errorf("updating UserProjectBinding: %w", err)
+	}
+
+	return nil
+}
+
+func (r *userProjectBindingReconciler) getGrafanaOrgUser(ctx context.Context, orgID uint, email string) (*grafanasdk.OrgUser, error) {
+	users, err := r.grafanaClient.GetOrgUsers(ctx, orgID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, user := range users {
+		if user.Email == email {
+			return &user, nil
+		}
+	}
+	return nil, nil
+}
+
+func (r *userProjectBindingReconciler) addGrafanaOrgUser(ctx context.Context, orgID uint, email, role string) (*grafanasdk.OrgUser, error) {
+	req, err := http.NewRequest("GET", r.grafanaURL+"/api/user", nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add(r.grafanaHeader, email)
+	resp, err := r.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	type response struct {
+		ID uint `json:"id"`
+	}
+	res := &response{}
+	defer resp.Body.Close()
+	decoder := json.NewDecoder(resp.Body)
+	if err := decoder.Decode(res); err != nil || res.ID == 0 {
+		return nil, fmt.Errorf("unable to decode response : %w", err)
+	}
+	// delete user from default org
+	if status, err := r.grafanaClient.DeleteOrgUser(ctx, defaultOrgID, res.ID); err != nil {
```

Comment:
I have few concerns about this implementation:
1. Where is the code to add users to the default organization? I didn't find that, and only see the code to remove the user from the default organization.
2. If this is failing all the time, does it mean that all users will be in the same default org? In that case, the user with `admin` role can kick other users (including the admin user that we are using in this controller) out of the default organization, which will lead to permission escalations.

Reply:
1. why do you need to add somebody to default org? default org - is the org where new auth users assigned by default, so we don't need to add anyone there, we only moving users from default org to the org where they should be. 
2. users that added to default org does not have admin right

Reply:
also, this method won't fail if user already removed from default org

Reply:
> also, this method won't fail if user already removed from default org

But I think it can fail due to some connection issues?

> users that added to default org does not have admin right

This is good, but still IMO it is not good to have all KKP users in one single default organization, because I think that will leak customer information to each other. They can log in to the default organization and see all users?

Actaully this behavior is configurable: https://grafana.com/docs/grafana/latest/administration/configuration/#auto_assign_org
You can disable the auto assignment.

Reply:
Oh, actually the user role in the default organization is also configurable: https://grafana.com/docs/grafana/latest/administration/configuration/#auto_assign_org_role
and it is `Viewer` by default, and for `Viewer`, they can not manage users: https://grafana.com/docs/grafana/latest/permissions/organization_roles/#compare-roles
Does this mean that they even cannot see users? If yes, I think we are good :)

Reply:
if it fail due to some connection issue, we need to fail and try on the next time

> Actaully this behavior is configurable: https://grafana.com/docs/grafana/latest/administration/configuration/#auto_assign_org
> You can disable the auto assignment.

have you read it? :)
>When set to false, new users automatically cause a new organization to be created for that new user.

Reply:
> have you read it? :)
Yes, of course, but if we can not avoid leaking information of customers to each other, do we have a better choice?

And as I mentioned, because I see that user will have `Viewer` by default, if they cannot see each other, then I think we can put them in the single default org.